mlir/lib/Dialect/MQTOpt/Transforms/Transpilation/sc/RoutingPass.cpp

```diff
@@ -0,0 +1,1071 @@
+            programIdx1, hardwareIdx0, programIdx0, hardwareIdx1, programIdx0,
+            hardwareIdx0, programIdx1, hardwareIdx1);
+      });
+
+      auto swap = createSwap(op->getLoc(), qIn0, qIn1, rewriter);
+      const auto [qOut0, qOut1] = getOuts(swap);
+
+      rewriter.setInsertionPointAfter(swap);
+      replaceAllUsesInRegionAndChildrenExcept(
+          qIn0, qOut1, swap->getParentRegion(), swap, rewriter);
+      replaceAllUsesInRegionAndChildrenExcept(
+          qIn1, qOut0, swap->getParentRegion(), swap, rewriter);
+
+      stack.recordSwap(programIdx0, programIdx1);
+
+      auto& state = stack.topState();
+      state.swapProgramIndices(qIn0, qIn1);
+      state.remapQubitValue(qIn0, qOut0);
+      state.remapQubitValue(qIn1, qOut1);
+
+      checkZeroUse(qOut0, stack, pool);
+      checkZeroUse(qOut1, stack, pool);
+    }
+  }
+};
+
+/**
+ * @brief The necessary datastructures to route a quantum-classical module.
+ */
+struct RoutingContext {
+  explicit RoutingContext(std::unique_ptr<Architecture> arch,
+                          std::unique_ptr<RouterBase> router,
+                          std::unique_ptr<InitialLayoutGeneratorBase> ilg)
+      : arch_(std::move(arch)), router_(std::move(router)),
+        ilg_(std::move(ilg)) {}
+
+  [[nodiscard]] Architecture& arch() const { return *arch_; }
+  [[nodiscard]] RouterBase& router() const { return *router_; }
+  [[nodiscard]] InitialLayoutGeneratorBase& ilg() const { return *ilg_; }
+  [[nodiscard]] StateStack& stack() { return stack_; }
+  [[nodiscard]] HardwareIndexPool& pool() { return pool_; }
+
+private:
+  std::unique_ptr<Architecture> arch_;
+  std::unique_ptr<RouterBase> router_;
+  std::unique_ptr<InitialLayoutGeneratorBase> ilg_;
+  StateStack stack_{};
+  HardwareIndexPool pool_{};
+};
+
+/**
+ * @brief Adds nqubit 'mqtopt.qubit' ops for entry_point functions. Initializes
+ * the pool and hence applies the initial layout. Consequently, the initial
+ * layout is always applied at the beginning of a function. Pushes
+ * newly-initialized state on stack.
+ */
+WalkResult handleFunc(func::FuncOp op, RoutingContext& ctx,
+                      PatternRewriter& rewriter) {
+  assert(ctx.stack().empty() && "handleFunc: stack must be empty");
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "handleFunc: entry_point= " << op.getSymName() << '\n';
+  });
+
+  ctx.ilg().generate();
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "handleFunc: initial layout= ";
+    ctx.ilg().dump();
+    llvm::dbgs() << '\n';
+  });
+
+  rewriter.setInsertionPointToStart(&op.getBody().front());
+
+  llvm::SmallVector<Value> qubits(ctx.arch().nqubits());
+  for (std::size_t i = 0; i < qubits.size(); ++i) {
+    auto qubitOp =
+        rewriter.create<QubitOp>(rewriter.getInsertionPoint()->getLoc(), i);
+    rewriter.setInsertionPointAfter(qubitOp);
+    qubits[i] = qubitOp.getQubit();
+  }
+
+  ctx.pool().fill(ctx.ilg().getLayout());
+  ctx.stack().emplace(LayoutState(qubits, ctx.ilg().getLayout()));
```

Comment:
```suggestion
  ctx.stack().emplace(qubits, ctx.ilg().getLayout());
```
if set up correctly, this should also work and avoid the construction of a temporary.